lib/db/backend/backend.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (C) 2019 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package backend
+
+import (
+	"sync"
+)
+
+type Reader interface {
+	Get(key []byte) ([]byte, error)
+	NewPrefixIterator(prefix []byte) (Iterator, error)
+	NewRangeIterator(first, last []byte) (Iterator, error)
+}
+
+type Writer interface {
+	Put(key, val []byte) error
+	Delete(key []byte) error
+}
+
+type ReadTransaction interface {
+	Reader
+	Release()
+}
+
+type WriteTransaction interface {
+	ReadTransaction
+	Writer
+	Commit() error
+}
+
+type Iterator interface {
+	Next() bool
+	Key() []byte
+	Value() []byte
+	Error() error
```

Comment:
How do I realize there is an error while iterating - presumably I'd need to call `Error` every time after `Next`? Looking at code sometimes that's done, sometimes not.  
I'd expect to have an error return value on `Next`, and then no error on `Value`/`Key` if that succeeds, as my expectation would be that calling `Next` already does look those values up. Otherwise I'd expect an error return value on all of them. That's a bit cumbersome probably, but most consistent with other methods and least surprising in my opinion.

Reply:
The underlying details vary... The idea with this interface is that you check it after the loop, that is if there's a read error then Next() will return false to stop the iteration and Error() will return why (as opposed to nil if the iteration just reached the end). Maybe I've missed to do that somewhere.

We could have Next() return an error directly instead, but then you need to compare that to some sentinel and the code becomes more cumbersome as you may want access to that error after the iteration so it cannot just become a `for Next() == nil` thing.

That Key() does not return an error is because there is no possibility for it to error once iteration to a key has succeeded. Value() *can* fail in at least the badger implementation because values are not always prefetched. So it's some abstraction leakage if you will.

Reply:
Got it. It seems non-ideal to put the onus on the consumer to remember that after `.Next() == False` one needs to check `.Error()` to be sure the iteration aborted successfully at the end. Maybe an alternative approach would be to let `Release` return the error, thus it's at least guaranteed the consumer gets into contact with an error.

Reply:
My point is simply that it (distinction end reached and aborted due to error) seems very easy to miss and unfortunately I can't think of any simple/clean way to make it more in your face. So probably just what Audrius wrote somewhere: Add some comment to the interface specifying the behaviour (`Next` returns false on error and when reaching end, and `Error` needs to be called afterwards to detect if an error occurred). In principle this can be inferred from the interface, but I believe it's unlikely to be immediately obvious in practice.

Reply:
Gotcha. I'm doing a revision soonish to handle the feedback and also some things that came out of the badger tests (where a transaction cannot be committed/discarded with active iterators, making the defers sometimes impractical).